src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+            padding_idx: int,
+    ):
+        super().__init__(num_embeddings, embedding_dim, padding_idx)
+        self.onnx_trace = False
+
+    def forward(self, input, use_cache=False, positions=None):
+        """Input is expected to be of size [bsz x seqlen]."""
+        assert (
+            (positions is None) or (self.padding_idx is None)
+        ), "If positions is pre-computed then padding_idx should not be set."
+
+        if positions is None:
+            if use_cache:
+                # positions is the same for every token when decoding a single step
+                # Without the int() cast, it doesn't work in some cases when exporting to ONNX
+                positions = input.data.new(1, 1).fill_(int(self.padding_idx + input.size(1)))
+            else:
+                mask = input.data.ne(self.padding_idx).int()
+                positions = ( torch.cumsum(mask, dim=1).type_as(mask) * mask ).long() + self.padding_idx
+            real_positions = positions
+        else:
+            real_positions = positions
+        return super().forward(positions), real_positions
+
+    def max_positions(self):
+        """Maximum number of supported positions."""
+        if self.padding_idx is not None:
+            return self.num_embeddings - self.padding_idx - 1
+        else:
+            return self.num_embeddings
+
+    def _forward(self, positions):
+        return super().forward(positions)
+
+def LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=True, export=False):
+    if not export and torch.cuda.is_available():
+        try:
+            from apex.normalization import FusedLayerNorm
+            return FusedLayerNorm(normalized_shape, eps, elementwise_affine)
+        except ImportError:
+            pass
+    return torch.nn.LayerNorm(normalized_shape, eps, elementwise_affine)
+
+def invert_mask(attention_mask):
+    assert attention_mask.dim() == 2
+    return attention_mask.eq(0)
+
+
+class SelfAttention(nn.Module):
+    """Multi-headed attention from 'Attention Is All You Need' paper"""
+
+    def __init__(
+        self,
+        embed_dim,
+        num_heads,
+        dropout=0.0,
+        bias=True,
+        encoder_decoder_attention=False,
+    ):
+        super().__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.head_dim = embed_dim // num_heads
+        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
+        self.scaling = self.head_dim ** -0.5
+
+        self.encoder_decoder_attention = encoder_decoder_attention
+        self.k_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.v_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.q_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.out_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.cache_key = "encoder_decoder" if self.encoder_decoder_attention else "self"
```

Comment:
this looks like complicated logic in the following lines :D Maybe add a comment?